### ISSUE TITLE
Environment variables are not updated

### ISSUE THREAD

**Issue by abellion** [parcel-bundler/parcel#1625]
# 🐛 bug report

I'm using dotenv to load environment variables from a `.env` file ; that's convenient to change their value quickly. Though, right now I always need to restart parcel in order to have them changed inside the bundle.

## 🎛 Configuration (.babelrc, package.json, cli command)

`parcel watch server/index.js -d dist/server --target=node`

## 🤔 Expected Behavior

The bundle should contains updated environment variables.

## 😯 Current Behavior

Only environment variables from when parcel was launched are alive.

## 💁 Possible Solution

Maybe don't cache env vars ?

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | latest of 1.8.x
| Node             | 8 LTS
| npm/Yarn         | latest npm
| Operating System | `Linux ce3280140caf 4.9.87-linuxkit-aufs #1 SMP Wed Mar 14 15:12:16 UTC 2018 x86_64 GNU/Linux`

**Comment by DeMoorJasper** [comment]
We should probably add the env variables to the cache key

**Comment by muratg** [comment]
Any updates? Also, any workarounds? I can manually clear the cache, but I'd rather not.

**Comment by BrOrlandi** [comment]
I want to know workaround too. It's impossible to change env vars here, even removing the cache, the variables still cached! This bug is really annoying and I don't know how to fix it. 😞

**Comment by cyrilchapon** [comment]
Problem is also appearing when targetting browser with [babel environment variables plugin](https://www.npmjs.com/package/babel-plugin-transform-inline-environment-variables).

**Comment by mischnic** [comment]
Does parcel rebuild when an env file is changed (and the env variables don't update) or does just nothing happen on edit?

**Comment by cyrilchapon** [comment]
Nothing happens, furthermore, if ctrl+c the parcel again, the cache is restored and the new env var values are not reflected in the code

**Comment by mischnic** [comment]
> if ctrl+c the parcel again, the cache is restored and the new env var values are not reflected in the code

That part is working for me.

**Comment by cyrilchapon** [comment]
Hum. Not for me.

**index.html**
```html
<script src="./index.js"></script>
```

**index.js**
```js
console.log(process.env.FOO)
```

**cmd**
```shell
> rm -rf .parcel-cache dist
> FOO=bar parcel build src/index.html --cache-dir .parcel-cache
✨  Built in 23.04s.

dist/src.050c5e28.map        ⚠️  5.67 MB     338ms
dist/src.050c5e28.js          ⚠️  1.9 MB    22.16s
dist/src.6a665fea.css          23.09 KB     6.17s
dist/favicon.8d872578.ico      16.56 KB     307ms
dist/index.html                   535 B     460ms
```

**result**
```
// bar
```

Now change it

**cmd**
```shell
cmd+c
> FOO=somthing parcel build src/index.html --cache-dir .parcel-cache
✨  Built in 1.91s.

dist/src.050c5e28.map        ⚠️  5.67 MB    376ms
dist/src.050c5e28.js          ⚠️  1.9 MB    1.36s
dist/src.6a665fea.css          23.09 KB     36ms
dist/favicon.8d872578.ico      16.56 KB     16ms
dist/index.html                   535 B      9ms
```

**Wrong result**
```
// bar
```

NB1: mind the fact that the digest in the asset slug name did not change after the rebuild
NB2: dotenv or inline environment variables changes nothing

**Comment by cyrilchapon** [comment]
As a workaround, for now, I consider these 2 facts:
- Anyway everyone is used to have to restart the server when they change an .env var (that's the case with webpack, nodemon, and any other common tool that provide server auto restart)
- Having the rebuild-from-cache feature is mostly safe and intersting during fast development iterations and **source code** updates

Thus, I have the following in package.json
```
    "uncache": "shx rm -rf .parcel-cache",
    "compile": "parcel build src/index.html --cache-dir .parcel-cache",
    "build": "npm run uncache && npm run compile",
    "uncache:dev": "shx rm -rf .parcel-dev-cache",
    "serve": "parcel src/index.dev.html --cache-dir .parcel-dev-cache",
    "dev": "npm run uncache:dev && npm run serve",
```

Which provides :
- in production, cache cleanup before each build (as a safety condition)
- in dev, hot reload with rebuild-from-cache when editing sources, but cache-cleanup between each dev server cold start

The only limitation, which anyway everyone is used to, is that you have to cold restart manually the server (killing the process) when you have to change an environment variable.

**Though**, there is still a bug. The fact I have to manually erase the cache before a cold start should not happen, and as @DeMoorJasper said
> We should probably add the env variables to the cache key

**Comment by cyrilchapon** [comment]
I may take an hour to contribute on this, @DeMoorJasper @mischnic can you point me to the right direction to "add that environment change to the cache key ?"

**Comment by mischnic** [comment]
> EDIT: I found this old issue #66
> This seems like a regression then ?

Yes, this function should make caching work properly for env variables:
https://github.com/parcel-bundler/parcel/blob/30624f75b7421c0627202e459240d68e1b50bbbb/packages/core/parcel-bundler/src/utils/isAccessedVarChanged.js#L4-L12

**Comment by mischnic** [comment]
Both examples from this comment https://github.com/parcel-bundler/parcel/issues/1625#issuecomment-470002629 works for me.
Which version of Parcel are you using? Could you try it with the current master?

**Comment by cyrilchapon** [comment]
That's curious

`"parcel-bundler": "^1.11.0"`

I'll try and come back

**Comment by cyrilchapon** [comment]
Well I'm back

> Could you try it with the current master?

Nope I can't. That's what happen when a repo is only-yarn-not-npm compatible.

Just

```
git clone [stuff] && cd parcel
npm install
```

doesn't work
```
git clone [stuff] && cd parcel
yarn install
cd packages/core/parcel-bundler
npm link
```
Doesn't work

**Comment by ChristophP** [comment]
Any updates on this? I am experiencing the same behavior.
```sh
echo "FOO=bar" > .env
echo "<script>console.log(process.env.FOO)</script>" > index.html
npx parcel index.html # results in <script>console.log("bar")</script>

echo "FOO=BAR" > .env
npx parcel index.html # still results in <script>console.log("bar")</script> and not "BAR"

restarting parcel does NOT reflect any new changes.
clearing the cache makes it work, but that is annoying to do.

```

**Comment by muratg** [comment]
@ChristophP the only other "workaround" I found was using `--no-cache`. Still fast enough in most cases, but for my current project, I just went with create-react-app/webpack. I may give Parcel a try again once this is fixed because I don't enjoy webpack configuration at all.

**Comment by ChristophP** [comment]
@muratg yeah, the env thing was not a deal breaker for me but I just ran into another issue about broken source maps in Chrome which I think is. I also wanna use parcel so bad because I don't want to have to deal with webpack again. Desperately waiting for parcel 2 to get out of alpha :-)

**Comment by cmditch** [comment]
Experiencing the same error as you @cyrilchapon. I'll change a value in `.env`, run `parcel build ...`, and the new value shows up nowhere. Deleting the cache is the only fix.

(@ChristophP also experiencing issues with source maps in Chrome as well)

**Comment by abellion** [comment]
Bot : I didn't try on V2, this might not be resolved.

**Comment by ajm113** [comment]
Same issue for me. Annoying since I have to basically nuke my `.cache` folder every time just to deploy to 3-4 different sites just because env variables don't properly update when even running `parcel build`

**Comment by alex-klock** [comment]
In case its help to anyone having similar issues, we had to update our babel.config and specifically tell `transform-inline-environment-variables` to exclude which env vars we didn't want to be inlined (and therefor cached).

```
  plugins: [
    ['transform-inline-environment-variables', {
      'exclude': ['VARIABLE_ONE', 'VARIABLE_TWO'],
    }],
  ]
```

**Comment by ajm113** [comment]
@alex-klock Thanks for the tip! :) I tried it so far it actually seems to be working. Still need to do a little bit more tests, but so far it seems to correct the issue.

**Comment by EyMaddis** [comment]
how can a plugin (parcel 2) decide that it adds a cache dependency? 
I have a custom resolver that chooses which file to include based on an environment variable. Using the process.env.SOMETHING within a file works, but what if the path to the file changes, how can my Plugin invalidate that?

**Comment by mischnic** [comment]
@EyMaddis That's a good point, at the moment you can only invalidate based on file creations/modifications

**Comment by EyMaddis** [comment]
And what about transformers? Can I maybe implement a transformer that goes along with the resolver that can invalidate the cache? (If so, how?)

**Comment by mischnic** [comment]
Transformers can call `asset.invalidateOnEnvChange("SOMETHING")`: https://github.com/parcel-bundler/parcel/blob/a3679960728140a609ff29d98bf069e15d773305/packages/core/types/index.js#L580

but the resolver cache doesn't have anything to do with transformers.

**Comment by pulpdood** [comment]
I was pulling my hairs out on this. 

I tried clearing .parcel.cache and dist/bin folders and the environment variable was still picking up an old value. I just wanted to share my predicament, that I had the environment variable within my shell (`echo $ENV_VAR_IN_QUESTION` showed it was existing in my shell) and it was overwriting the environment variable in my `.env.development`. I just had to do `unset ENV_VAR_IN_QUESTION` in the shell, and got the updated environment variable.

I'm sorry if this is a bit off-topic, but I just wanted to help anyone who would have had the same predicament as me and help them.
